dockerclient.go

```diff
@@ -212,6 +212,50 @@ func (client *DockerClient) ContainerChanges(id string) ([]*ContainerChanges, er
 	return changes, nil
 }
 
+func (client *DockerClient) readJSONStream(stream io.ReadCloser, decode func(*json.Decoder) decodingResult) (<-chan decodingResult, chan<- struct{}) {
+	resultChan := make(chan decodingResult)
+	closeChan := make(chan struct{})
+	go func() {
+		defer close(resultChan)
+
+		internalResultsChan := make(chan decodingResult)
+		defer close(internalResultsChan)
+
+		stillListening := make(chan struct{})
+		defer close(stillListening)
+
+		go func() {
```

Comment:
what specifically? I can give a bit of an overview: this nested routine reads JSON from the stream and, if the other routine is still listening to `internalResultsChan` (this is the `stillListening` check) sends the result to the `internalResultsChan`. If the other routine is done or we read an error from the stream, then we stop reading from the stream and return.